compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -4572,6 +4606,262 @@ public void visit(BLangTypedescExpr accessExpr) {
+                (referredType.tag == TypeTags.FINITE || referredType.tag == TypeTags.UNION)) {
+            if (referredType.tag == TypeTags.FINITE) {
+                actualType = createFiniteTypeForNumericUnaryExpr(unaryExpr);
+            } else {
+                if (checkForCompatibleFiniteMembersInUnionType(unaryExpr, (BUnionType) referredType)) {
+                    return createFiniteTypeForNumericUnaryExpr(unaryExpr);
+                }
+                // Check for int subtypes
+                LinkedHashSet<BType> intTypesInUnion = getIntSubtypesInUnionType(unaryExpr, (BUnionType) referredType);
+                if (!intTypesInUnion.isEmpty()) {
+                    BType newReferredType = BUnionType.create(null, intTypesInUnion);
+                    BType tempActualType = checkCompatibilityWithConstructedNumericLiteral(unaryExpr, newReferredType);
+                    if (tempActualType != symTable.semanticError) {
+                        return  tempActualType;
+                    }
+                }
+            }
+        } else if (exprType != symTable.semanticError && TypeTags.isIntegerTypeTag(referredType.tag) &&
+                referredType.tag != TypeTags.INT) {
+            BType tempActualType = checkCompatibilityWithConstructedNumericLiteral(unaryExpr, referredType);
+            if (tempActualType != symTable.semanticError) {
+                return  tempActualType;
+            }
+        }
+        return actualType;
+    }
+
+    public BType checkCompatibilityWithConstructedNumericLiteral(BLangUnaryExpr unaryExpr, BType referredType) {
+        BLangNumericLiteral numericLiteral = createNumericLiteralFormUnaryExpr(unaryExpr);
+        // To check value with sign against expected type
+        return clonedExprTypeChecker(numericLiteral, referredType);
+    }
+
+    public LinkedHashSet getIntSubtypesInUnionType(BLangUnaryExpr unaryExpr, BUnionType expectedType) {
+        LinkedHashSet<BType> intTypesInUnion = new LinkedHashSet<>();
+        for (BType type : expectedType.getMemberTypes()) {
+            BType referredType = types.getReferredType(type);
+            if (TypeTags.isIntegerTypeTag(referredType.tag) && type.tag != TypeTags.INT) {
+                intTypesInUnion.add(referredType);
+            }
+        }
+        return intTypesInUnion;
+    }
+
+    public boolean checkForCompatibleFiniteMembersInUnionType(BLangUnaryExpr unaryExpr, BUnionType expectedType) {
+        BType compatibleTypeOfUnaryExpression;
+        for (BType type : expectedType.getMemberTypes()) {
+            compatibleTypeOfUnaryExpression = clonedExprTypeChecker(unaryExpr, types.getReferredType(type));
+            if (compatibleTypeOfUnaryExpression.tag == TypeTags.FINITE) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public BType clonedExprTypeChecker(BLangExpression expr, BType referredType) {
```

Comment:
Something like this is better IMO

```suggestion
    public BType checkClonedExprSilently(BLangExpression expr, BType referredType) {
```